homeassistant/components/bmw_connected_drive/__init__.py

```diff
@@ -44,49 +61,148 @@
-        discovery.load_platform(hass, component, DOMAIN, {}, config)
+    hass.async_create_task(
+        discovery.async_load_platform(hass, NOTIFY_DOMAIN, DOMAIN, {}, entry.data)
+    )
 
     return True
 
 
-def setup_account(account_config: dict, hass, name: str) -> "BMWConnectedDriveAccount":
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in BMW_PLATFORMS
+                if component != NOTIFY_DOMAIN
+            ]
+        )
+    )
+
+    # Only remove services if it is the last account and not read only
+    if (
+        len(hass.data[DOMAIN]) == 1
+        and not hass.data[DOMAIN][entry.entry_id][CONF_ACCOUNT].read_only
+    ):
+        services = list(_SERVICE_MAP) + [SERVICE_UPDATE_STATE]
+        for service in services:
+            hass.services.async_remove(DOMAIN, service)
+
+    for vehicle in hass.data[DOMAIN][entry.entry_id][CONF_ACCOUNT].account.vehicles:
+        hass.services.async_remove(NOTIFY_DOMAIN, slugify(f"{DOMAIN}_{vehicle.name}"))
+
+    if unload_ok:
+        hass.data[DOMAIN][entry.entry_id][UNDO_UPDATE_LISTENER]()
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+async def update_listener(hass, config_entry):
+    """Handle options update."""
+    await hass.config_entries.async_reload(config_entry.entry_id)
+
+
+def setup_account(entry: ConfigEntry, hass, name: str) -> "BMWConnectedDriveAccount":
     """Set up a new BMWConnectedDriveAccount based on the config."""
-    username = account_config[CONF_USERNAME]
-    password = account_config[CONF_PASSWORD]
-    region = account_config[CONF_REGION]
-    read_only = account_config[CONF_READ_ONLY]
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    region = entry.data[CONF_REGION]
+    read_only = entry.options[CONF_READ_ONLY]
+    use_location = entry.options[CONF_USE_LOCATION]
 
     _LOGGER.debug("Adding new account %s", name)
-    cd_account = BMWConnectedDriveAccount(username, password, region, name, read_only)
 
-    def execute_service(call):
-        """Execute a service for a vehicle.
+    pos = (
+        (hass.config.latitude, hass.config.longitude) if use_location else (None, None)
+    )
+    cd_account = BMWConnectedDriveAccount(
+        username, password, region, name, read_only, *pos
+    )
 
-        This must be a member function as we need access to the cd_account
-        object here.
-        """
+    def execute_service(call):
+        """Execute a service for a vehicle."""
         vin = call.data[ATTR_VIN]
-        vehicle = cd_account.account.get_vehicle(vin)
+        vehicle = None
+        # Double check for read_only accounts as another account could create the services
+        for entity in filter(
```

Comment:
Please us a list comprehension instead of filter + lambda.

Reply:
The variable is not an entity.
```suggestion
        for entry_data in filter(
```